File: gateway/evals/skill_router_baseline_report.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import Any


ZERO_SHA = "0000000000000000000000000000000000000000"


def _normalize_optional_text(value: Any) -> str | None:
    if not isinstance(value, str):
        return None
    normalized = value.strip()
    if not normalized or normalized == ZERO_SHA:
        return None
    return normalized


def resolve_base_sha(*, event_name: str, pr_base_sha: Any, before_sha: Any) -> str | None:
    if event_name == "pull_request":
        return _normalize_optional_text(pr_base_sha)
    return _normalize_optional_text(before_sha)


def _write_github_output(*, github_output_path: Path, available: bool) -> None:
    with github_output_path.open("a", encoding="utf-8") as handle:
        handle.write(f"available={'true' if available else 'false'}\n")


def _run_subprocess(command: list[str], *, cwd: Path | None = None) -> subprocess.CompletedProcess[str]:
    return subprocess.run(
        command,
        cwd=str(cwd) if isinstance(cwd, Path) else None,
        capture_output=True,
        text=True,
        check=False,
    )


def build_skill_router_baseline_report(
    *,
    event_name: str,
    pr_base_sha: str | None,
    before_sha: str | None,
    repo_root: Path,
    output_path: Path,
    python_bin: str,
    policy_rel_path: str = "gateway/evals/skill_router_policy.ci.json",
    eval_rel_path: str = "gateway/evals/skill_router_eval.py",
) -> dict[str, Any]:
    base_sha = resolve_base_sha(
        event_name=event_name,
        pr_base_sha=pr_base_sha,
        before_sha=before_sha,
    )
    if base_sha is None:
        return {"available": False, "reason": "no_base_sha", "base_sha": None}

    worktree_dir = Path(tempfile.mkdtemp(prefix="grobot-skill-router-base-", dir="/tmp"))
    try:
        add_result = _run_subprocess(
            ["git", "worktree", "add", "--detach", str(worktree_dir), base_sha],
            cwd=repo_root,
        )
        if add_result.returncode != 0:
            return {
                "available": False,
                "reason": "worktree_add_failed",
                "base_sha": base_sha,
                "stderr": add_result.stderr.strip(),
            }

        eval_path = worktree_dir / eval_rel_path
        policy_path = worktree_dir / policy_rel_path
        if not eval_path.exists() or not policy_path.exists():
            return {
                "available": False,
                "reason": "required_files_missing",
                "base_sha": base_sha,
            }

        output_path.parent.mkdir(parents=True, exist_ok=True)
        run_result = _run_subprocess(
            [
                python_bin,
                str(eval_path),
                "--policy",
                str(policy_path),
                "--print-json",
                "--output",
                str(output_path),
            ],
            cwd=worktree_dir,
        )
        available = output_path.exists()
        return {
            "available": available,
            "reason": "output_present" if available else "output_missing",
            "base_sha": base_sha,
            "eval_returncode": run_result.returncode,
            "eval_stderr": run_result.stderr.strip(),
        }
    finally:
        _run_subprocess(["git", "worktree", "remove", "--force", str(worktree_dir)], cwd=repo_root)
        shutil.rmtree(worktree_dir, ignore_errors=True)


def _build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Build skill-router baseline report from base commit via git worktree."
    )
    parser.add_argument("--event-name", default="", help="GitHub event name")
    parser.add_argument("--pr-base-sha", default="", help="PR base SHA for pull_request events")
    parser.add_argument("--before-sha", default="", help="github.event.before SHA for push events")
    parser.add_argument("--repo-root", type=Path, default=Path("."), help="repository root path")
    parser.add_argument(
        "--output",
        type=Path,
        default=Path("gateway/evals/data/skill_router_ci_report.base.json"),
        help="output path for baseline report",
    )
    parser.add_argument(
        "--github-output",
        type=Path,
        default=None,
        help="GitHub output file path; writes available=true|false",
    )
    parser.add_argument(
        "--python-bin",
        default=sys.executable,
        help="python executable used to run skill_router_eval.py",
    )
    parser.add_argument("--print-json", action="store_true", help="print result JSON")
    return parser


def main() -> None:
    parser = _build_parser()
    args = parser.parse_args()

    try:
        result = build_skill_router_baseline_report(
            event_name=args.event_name,
            pr_base_sha=args.pr_base_sha,
            before_sha=args.before_sha,
            repo_root=args.repo_root.resolve(),
            output_path=args.output.resolve(),
            python_bin=args.python_bin,
        )
    except Exception as error:  # noqa: BLE001
        result = {
            "available": False,
            "reason": "runtime_error",
            "error": f"{type(error).__name__}: {error}",
        }

    if isinstance(args.github_output, Path):
        args.github_output.parent.mkdir(parents=True, exist_ok=True)
        _write_github_output(github_output_path=args.github_output, available=bool(result.get("available")))

    if args.print_json:
        print(json.dumps(result, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: gateway/src/governance/evals/skill-router-baseline-report.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { spawnSync } from "node:child_process";

const ZERO_SHA = "0000000000000000000000000000000000000000";
const TSX_PACKAGE = "tsx@4.20.6";

type JsonObject = Record<string, unknown>;

interface ParsedCliArgs {
  eventName: string;
  prBaseSha: string;
  beforeSha: string;
  repoRoot: string;
  outputPath: string;
  githubOutputPath: string | undefined;
  pythonBin: string;
  printJson: boolean;
}

function normalizeOptionalText(value: unknown): string | undefined {
  if (typeof value !== "string") {
    return undefined;
  }
  const normalized = value.trim();
  if (!normalized || normalized === ZERO_SHA) {
    return undefined;
  }
  return normalized;
}

export function resolveBaseSha(input: {
  eventName: string;
  prBaseSha: unknown;
  beforeSha: unknown;
}): string | undefined {
  if (input.eventName === "pull_request") {
    return normalizeOptionalText(input.prBaseSha);
  }
  return normalizeOptionalText(input.beforeSha);
}

function dirname(path: string): string {
  const normalized = path.replace(/[\\/]+$/, "");
  const slash = normalized.lastIndexOf("/");
  if (slash <= 0) {
    return ".";
  }
  return normalized.slice(0, slash);
}

function removeTrailingSlashes(value: string): string {
  return value.replace(/[\\/]+$/, "");
}

function pathJoin(base: string, relative: string): string {
  const trimmedBase = removeTrailingSlashes(base);
  const trimmedRelative = relative.replace(/^[\\/]+/, "");
  return `${trimmedBase}/${trimmedRelative}`;
}

function isAbsolutePath(path: string): boolean {
  return path.startsWith("/");
}

function resolvePathFromRepoRoot(repoRoot: string, path: string): string {
  if (isAbsolutePath(path)) {
    return path;
  }
  return pathJoin(repoRoot, path);
}

function runCapture(command: string[]): {
  exitCode: number;
  stdout: string;
  stderr: string;
} {
  if (command.length === 0) {
    return { exitCode: 1, stdout: "", stderr: "empty command" };
  }
  const result = spawnSync(command[0], command.slice(1), {
    encoding: "utf8",
    maxBuffer: 8 * 1024 * 1024,
    timeout: 120_000,
  });
  const status = typeof result.status === "number" ? result.status : 1;
  const stdout = typeof result.stdout === "string" ? result.stdout : "";
  const stderr = typeof result.stderr === "string" ? result.stderr : "";
  return { exitCode: status, stdout, stderr };
}

function writeGithubOutput(path: string, available: boolean): void {
  mkdirSync(dirname(path), { recursive: true });
  const previous = existsSync(path) ? readFileSync(path, "utf8") : "";
  writeFileSync(path, `${previous}available=${available ? "true" : "false"}\n`, "utf8");
}

function parseArgs(argv: string[]): ParsedCliArgs {
  let eventName = "";
  let prBaseSha = "";
  let beforeSha = "";
  let repoRoot = ".";
  let outputPath = "gateway/evals/data/skill_router_ci_report.base.json";
  let githubOutputPath: string | undefined;
  let pythonBin = "python3";
  let printJson = false;

  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    if (token === "--event-name") {
      eventName = argv[index + 1] ?? "";
      index += 1;
      continue;
    }
    if (token === "--pr-base-sha") {
      prBaseSha = argv[index + 1] ?? "";
      index += 1;
      continue;
    }
    if (token === "--before-sha") {
      beforeSha = argv[index + 1] ?? "";
      index += 1;
      continue;
    }
    if (token === "--repo-root") {
      repoRoot = argv[index + 1] ?? ".";
      index += 1;
      continue;
    }
    if (token === "--output") {
      outputPath = argv[index + 1] ?? outputPath;
      index += 1;
      continue;
    }
    if (token === "--github-output") {
      githubOutputPath = argv[index + 1] ?? "";
      index += 1;
      continue;
    }
    if (token === "--python-bin") {
      pythonBin = argv[index + 1] ?? pythonBin;
      index += 1;
      continue;
    }
    if (token === "--print-json") {
      printJson = true;
      continue;
    }
  }

  return {
    eventName,
    prBaseSha,
    beforeSha,
    repoRoot,
    outputPath,
    githubOutputPath,
    pythonBin,
    printJson,
  };
}

interface BuildBaselineInput {
  eventName: string;
  prBaseSha: string;
  beforeSha: string;
  repoRoot: string;
  outputPath: string;
  pythonBin: string;
  policyRelPath?: string;
  evalRelPath?: string;
}

function randomSuffix(): string {
  return `${Date.now()}-${Math.abs(Math.trunc(Math.random() * 1_000_000))}`;
}

export function buildSkillRouterBaselineReport(input: BuildBaselineInput): JsonObject {
  const baseSha = resolveBaseSha({
    eventName: input.eventName,
    prBaseSha: input.prBaseSha,
    beforeSha: input.beforeSha,
  });
  if (!baseSha) {
    return { available: false, reason: "no_base_sha", base_sha: null };
  }

  const repoRoot = removeTrailingSlashes(input.repoRoot);
  const outputPath = resolvePathFromRepoRoot(repoRoot, input.outputPath);
  const policyRelPath = input.policyRelPath ?? "gateway/evals/skill_router_policy.ci.json";
  const evalRelPath = input.evalRelPath ?? "gateway/src/governance/evals/skill-router-eval.ts";
  const worktreeDir = `/tmp/grobot-skill-router-base-${randomSuffix()}`;

  const addResult = runCapture(["git", "-C", repoRoot, "worktree", "add", "--detach", worktreeDir, baseSha]);
  if (addResult.exitCode !== 0) {
    return {
      available: false,
      reason: "worktree_add_failed",
      base_sha: baseSha,
      stderr: addResult.stderr.trim(),
    };
  }

    try {
      const evalPath = pathJoin(worktreeDir, evalRelPath);
      const policyPath = pathJoin(worktreeDir, policyRelPath);
      if (!existsSync(evalPath) || !existsSync(policyPath)) {
        return {
          available: false,
        reason: "required_files_missing",
        base_sha: baseSha,
      };
    }

    mkdirSync(dirname(outputPath), { recursive: true });
    const evalCommand: string[] = [];
    const normalizedEvalPath = evalPath.toLowerCase();
    if (normalizedEvalPath.endsWith(".ts")) {
      evalCommand.push("npx", "--yes", "--package", TSX_PACKAGE, "tsx", evalPath);
    } else if (normalizedEvalPath.endsWith(".js")) {
      evalCommand.push("node", evalPath);
    } else {
      evalCommand.push(input.pythonBin, evalPath);
    }
    evalCommand.push("--policy", policyPath, "--print-json", "--output", outputPath);
    const runResult = runCapture(evalCommand);
    const available = existsSync(outputPath);
    return {
      available,
      reason: available ? "output_present" : "output_missing",
      base_sha: baseSha,
      eval_returncode: runResult.exitCode,
      eval_stderr: runResult.stderr.trim(),
    };
  } finally {
    runCapture(["git", "-C", repoRoot, "worktree", "remove", "--force", worktreeDir]);
  }
}

function main(): number {
  const args = parseArgs(process.argv.slice(2));
  let result: JsonObject;
  try {
    result = buildSkillRouterBaselineReport({
      eventName: args.eventName,
      prBaseSha: args.prBaseSha,
      beforeSha: args.beforeSha,
      repoRoot: args.repoRoot,
      outputPath: args.outputPath,
      pythonBin: args.pythonBin,
    });
  } catch (error) {
    result = {
      available: false,
      reason: "runtime_error",
      error: String(error),
    };
  }

  if (args.githubOutputPath && args.githubOutputPath.trim().length > 0) {
    writeGithubOutput(args.githubOutputPath.trim(), result.available === true);
  }
  if (args.printJson) {
    process.stdout.write(`${JSON.stringify(result, undefined, 0)}\n`);
  }
  return 0;
}

function isDirectExecution(): boolean {
  const entry = process.argv[1];
  if (typeof entry !== "string") {
    return true;
  }
  const normalized = entry.replace(/\\/g, "/");
  return (
    normalized.endsWith("/skill-router-baseline-report.ts") ||
    normalized.endsWith("/skill-router-baseline-report.js")
  );
}

if (isDirectExecution()) {
  process.exitCode = main();
}
